File: src/programmers/PRO_L3_예산.cpp
/* 예산
 * 이진탐색
 * 2020.01.12
 */
#include <algorithm>
#include <vector>

using namespace std;

long sumOfBudget(vector<int> budgets, int max_b) {

    long sum = 0L;
    for (auto b: budgets) {
        if (b <= max_b) sum += b;
        else sum += max_b;
    }
    return sum;
}

int solution(vector<int> budgets, int M) {

    sort(budgets.begin(), budgets.end());
    if (sumOfBudget(budgets, budgets.back()) <= M) {
        return budgets.back();
    } 

    // 탐색 범위: 0 ~ 국가 예산 중 가장 큰 값
    int low = 0;
    int high = budgets.back();
    int answer = 0;
    long temp = 0L;

    // mid: 모든 요청이 배정될 수 있는 상한액
    while (low <= high) {
        int mid = (low+high)/2;
        long target = sumOfBudget(budgets, mid);

        if (target == M) {
            return mid;
        } else if (target < M) {
            low = mid+1;
            // 가능한 상한액을 임시 저장
            if (temp < target) {
                temp = target;
                answer = mid;
            }
        } else {
            high = mid-1;
        }
    }
    return answer;
}

File: src/programmers/PRO_L3_예산.java
package programmers;

/**
 * @{link} 	https://programmers.co.kr/learn/courses/30/lessons/43237
 * @date   	2020-05-09
 * @author 	rkddlsgur983
 * @idea	이진탐색
 */
public class PRO_L3_예산 {
	public int solution(int[] budgets, int M) {
		int len = budgets.length;
		int min = 1;
		int max = len;
		for (int i = 0; i < len; ++i) {
			max = Math.max(budgets[i], max);
		}
        int answer = 0;
		while (min <= max) {
			int mid = (min+max)/2;
			boolean go = true;
			int sum = 0;
			for (int i = 0; i < len; ++i) {
				if (budgets[i] > mid) {
					sum += mid;
				} else {
					sum += budgets[i];
				}
				if (sum > M) {
					go = false;
					break;
				}
			}
			if (go) {
				answer = mid;
				min = mid+1;
			} else {
				max = mid-1;
			}
		}
        return answer;
    }
}
